geoengine/layers.py

```diff
@@ -0,0 +1,660 @@
+'''
+A wrapper around the layer and layerDb API
+'''
+
+from __future__ import annotations
+from dataclasses import dataclass
+from enum import auto
+from io import StringIO
+import os
+from typing import Any, Dict, List, NewType, Optional, Union, cast
+from uuid import UUID
+import json
+from typing_extensions import TypedDict
+import requests as req
+from strenum import LowercaseStrEnum
+from geoengine.auth import get_session
+from geoengine.error import GeoEngineException, ModificationNotOnLayerDbException
+
+
+LayerId = NewType('LayerId', UUID)
+LayerCollectionId = NewType('LayerCollectionId', UUID)
+LayerProviderId = NewType('LayerProviderId', UUID)
+
+LAYER_DB_PROVIDER_ID = LayerProviderId(UUID('ce5e84db-cbf9-48a2-9a32-d4b7cc56ea74'))
+LAYER_DB_ROOT_COLLECTION_ID = LayerCollectionId(UUID('05102bb3-a855-4a37-8a8a-30026a91fef1'))
+
+
+class LayerCollectionAndProviderIdResponse(TypedDict):
+    collectionId: str
+    providerId: str
+
+
+class LayerAndProviderIdResponse(TypedDict):
+    layerId: str
+    providerId: str
+
+
+class LayerCollectionResponse(TypedDict):
+    '''A layer collection response JSON from a HTTP request'''
+    id: LayerCollectionAndProviderIdResponse
+    name: str
+    description: str
+    items: List[LayerCollectionListingResponse]
+
+
+class LayerCollectionListingResponse(TypedDict):
+    '''A layer collection listing response JSON from a HTTP request'''
+    id: Union[LayerCollectionAndProviderIdResponse, LayerAndProviderIdResponse]
+    name: str
+    description: str
+    type: str
+
+
+class LayerCollectionListingType(LowercaseStrEnum):
+    LAYER = auto()
+    COLLECTION = auto()
+
+
+@dataclass(repr=False)
+class LayerCollectionListing:
+    '''A layer collection listing as item of a collection'''
+
+    listing_id: Union[LayerId, LayerCollectionId]
+    provider_id: LayerProviderId
+    name: str
+    description: str
+    type: LayerCollectionListingType
```

Comment:
why not use a subclass for each type? That would also help to get rid of any `if type == ...` statements in the code

Reply:
They are no subclasses. You can make isinstance in a later version of Python, when newtypes are actual types.

Reply:
There is inheritance aka subclasses in python or what am I missing? https://www.w3schools.com/python/python_inheritance.asp